Cache cargo dependencies in Docker layers for faster build

oliverbarnes opened the pull request:
Using Luca's cargo-chef: https://lpalmieri.com/posts/fast-rust-docker-builds/

Closes #66

marcoow commented:
I don't think I know enough about Docker to properly review this

LukeMathWalker commented on Dockerfile:
```diff
@@ -1,6 +1,28 @@
+
+FROM chef AS planner
+
+COPY ./Cargo.lock .
+COPY ./Cargo.toml .
+COPY ./cli ./cli
+COPY ./config ./config
+COPY ./macros ./macros
+COPY ./rust-toolchain.toml .
+COPY ./web ./web
+COPY ./db ./db
+COPY ./stations.sqlite.db ./stations.sqlite.db
```
```suggestion
COPY . .
```

No need to make it more complex than this, it won't affect dependency caching (since that depends exclusively on the has of `recipe.json`).

LukeMathWalker commented on Dockerfile:
```diff
@@ -1,6 +1,28 @@
-FROM rust:1.85 AS builder
+FROM lukemathwalker/cargo-chef:latest-rust-1.85.1-slim AS chef
 
 WORKDIR /usr/src/restations-builder
+
```
```suggestion
FROM lukemathwalker/cargo-chef:latest-rust-1.85.1-slim AS chef
WORKDIR /usr/src/restations-builder
COPY rust-toolchain.toml rust-toolchain.toml
# Make sure we sync the active toolchain once, if needed.
RUN rustup toolchain install
```

LukeMathWalker commented on Dockerfile:
```diff
@@ -1,6 +1,28 @@
+
+# Build dependencies - this is the caching Docker layer!
+RUN cargo chef cook --release --recipe-path recipe.json
+
```
Same down here, just go `COPY . .`.
If you need to ignore some files, using `.dockerignore`.

LukeMathWalker commented:
A few more issues:

- Don't use a Rust image as runtime image. Use debian or another similar "bare" image https://github.com/mainmatter/reStations/blob/fefc166644ca0b444df95c78ae922b6c21157381/Dockerfile#L27
- All `APP_*` (except `APP_ENVIRONMENT`) should really be handled via a production.yml configuration file https://github.com/mainmatter/reStations/blob/fefc166644ca0b444df95c78ae922b6c21157381/Dockerfile#L37

oliverbarnes commented:
> * Don't use a Rust image as runtime image. Use debian or another similar "bare" image https://github.com/mainmatter/reStations/blob/fefc166644ca0b444df95c78ae922b6c21157381/Dockerfile#L27

Changed to use debian:bookworm-slim

> * All `APP_*` (except `APP_ENVIRONMENT`) config values should really be handled via a production.yml configuration file https://github.com/mainmatter/reStations/blob/fefc166644ca0b444df95c78ae922b6c21157381/Dockerfile#L37

Moved this to a separate issue as it's a little more involved (for me at least)

> * I may be wrong, but all that stuff around user permissions should only happen for the runtime image.

Moved it under runtime. Not tested locally yet (I'm editing on Github)
